File: duckgo/groq/ai_client.py
import os
import re
from dotenv import load_dotenv
from groq import Groq
from rich.console import Console
from rich.markdown import Markdown
from duckgo.tools import WRITE_TOOL, RUN_COMMAND_TOOL, run_command, write_file
import json

load_dotenv(".env")

TOOLS = [WRITE_TOOL, RUN_COMMAND_TOOL]
TOOL_FUNCTIONS = {
    "write_file": write_file,
    "run_command": run_command,
    }

class AIClient:
    def __init__(self, api_key: str, model: str = "qwen/qwen3-32b"):
        self.client = Groq(api_key=api_key)
        self.model = model
        self.history: list[dict] = []
        
    def ask(self, prompt: str, system: str | None = None) -> str:
        if system and not self.history:
            self.history.append({"role": "system", "content": system})

        self.history.append({"role": "user", "content": prompt})

        max_iterations = 15
        for _ in range(max_iterations):
            response = self.client.chat.completions.create(
                model=self.model,
                messages=self.history,
                tools=TOOLS,
                tool_choice="auto",
            )

            message = response.choices[0].message

            if not message.tool_calls:
                reply = message.content
                self.history.append({"role": "assistant", "content": reply})
                return reply

            self.history.append(message)

            for call in message.tool_calls:
                args = json.loads(call.function.arguments)
                func = TOOL_FUNCTIONS[call.function.name]
                result = func(**args)

                self.history.append({
                    "role": "tool",
                    "tool_call_id": call.id,
                    "content": result,
                })

        return "Limite de iterações atingido — a tarefa pode ser complexa demais para uma única resposta."
    @staticmethod
    def _strip_thinking(text: str) -> str:
        return re.sub(r"<think>.*?</think>", "", text, flags=re.DOTALL).strip()

File: src/core/AIClient.ts
import type { AIProvider } from "../providers/Provider";
import type { Message, ToolDefinition } from "./types";

type ToolFunction = (args: Record<string, any>) => string | Promise<string>;

export class AIClient {
  private history: Message[] = [];
  private readonly maxIterations = 15;

  constructor(
    private provider: AIProvider,
    private model: string,
    private tools: ToolDefinition[],
    private toolFunctions: Record<string, ToolFunction>
  ) {}

  setSystemPrompt(prompt: string) {
    if (this.history.length === 0) {
      this.history.push({ role: "system", content: prompt });
    }
  }

  async ask(prompt: string): Promise<string> {
    this.history.push({ role: "user", content: prompt });

    for (let i = 0; i < this.maxIterations; i++) {
      const result = await this.provider.complete(this.history, this.tools, this.model);

      if (result.toolCalls.length === 0) {
        this.history.push({ role: "assistant", content: result.content });
        return result.content ?? "";
      }

      this.history.push({
        role: "assistant",
        content: result.content,
        toolCalls: result.toolCalls,
      });

      for (const call of result.toolCalls) {
        const func = this.toolFunctions[call.name];
        const toolResult = func ? await func(call.arguments) : "Ferramenta não encontrada.";

        this.history.push({
          role: "tool",
          content: toolResult,
          toolCallId: call.id,
        });
      }
    }

    return "Limite de iterações atingido.";
  }

  clearHistory() {
    const systemMsg = this.history.find((m) => m.role === "system");
    this.history = systemMsg ? [systemMsg] : [];
  }
}